### PR TITLE
drivers: uart-emul: Support for async API

### PR DESCRIPTION
**Pull request by rvoell-gcx**
Add support for the async uart API, so the uart emulation driver can now be used to test code using that API.

### CODE REVIEW

**Comment by dcpleung** [comment]
Please fix the compliance check errors.

**Comment by rvoell-gcx** [comment]
> Please fix the compliance check errors.

Yes, i'm already working on it

File: tests/drivers/uart/uart_emul/prj.conf [hunk 1]
```diff
@@ -1,3 +1,5 @@
 CONFIG_ZTEST=y
+CONFIG_EVENTS=y
 CONFIG_SERIAL=y
 CONFIG_UART_INTERRUPT_DRIVEN=y
+CONFIG_UART_ASYNC_API=y
```
Comment (dcpleung) [inline tests/drivers/uart/uart_emul/prj.conf:5]: It would be better to create another test in testcase.yaml so both async and non-async APIs are tested.

File: tests/drivers/uart/uart_async_api/src/test_uart.h [hunk 2]
```diff
@@ -27,4 +27,9 @@
 #define UART_NODE DT_CHOSEN(zephyr_console)
 #endif
 
+#if DT_NODE_HAS_COMPAT(UART_NODE, zephyr_uart_emul)
+#include <zephyr/drivers/serial/uart_emul.h>
+#define DUT_IS_UART_EMUL 1
+#endif
```
Comment (dcpleung) [inline tests/drivers/uart/uart_async_api/src/test_uart.h:33]: This doesn't seem right. It is possible that there is emulated UART, but the test is actually running with a real UART on hardware.

Comment (rvoell-gcx) [reply tests/drivers/uart/uart_async_api/src/test_uart.h:33]: Thats why i checked if `UART_NODE`, which is the one used in the testcases, is the emulated uart.

File: tests/drivers/uart/uart_emul/prj.conf (hunk 1 above)
Comment (rvoell-gcx) [reply tests/drivers/uart/uart_emul/prj.conf:5]: It just enables the async API. As far as i understand it, the APIs can coexist. Otherwise the test cases would fail, as each test case checks one of the APIs. Or am i missing something?

File: tests/drivers/uart/uart_async_api/src/test_uart_async.c [hunk 3]
```diff
@@ -66,6 +66,10 @@ static void uart_async_test_init(void)
 	k_sem_reset(&rx_buf_coherency);
 	k_sem_reset(&rx_buf_released);
 	k_sem_reset(&rx_disabled);
+#ifdef DUT_IS_UART_EMUL
+	uart_emul_flush_rx_data(uart_dev);
+	uart_emul_flush_tx_data(uart_dev);
+#endif /* DUT_IS_UART_EMUL */
```
Comment (bjarki-trackunit) [inline tests/drivers/uart/uart_async_api/src/test_uart_async.c:72]: The emulated driver should behave like a real uart, I don't expect any changes should be needed for the uart test suite. as this is a clear indication that the emulated uart is not behaving as expected :)

On real hardware, the following calls should clear the rx and tx data respectively
```
	uart_rx_disable(uart_dev);
	uart_tx_abort(uart_dev);
```

File: tests/drivers/uart/uart_async_api/src/test_uart_async.c [hunk 4]
```diff
@@ -584,7 +588,16 @@ ZTEST_USER(uart_async_read_abort, test_read_abort)
...
+#ifndef DUT_IS_UART_EMUL
 	k_timer_start(&read_abort_timer, K_USEC(300), K_NO_WAIT);
+#else
+	/* The uart emulation driver needs no real time to send and receive data. Any waiting here
+	 * will lead to the driver being finished with sending and receiving. So instead of using
+	 * the timer here, just disable receiving before the driver had any chance to start with
+	 * processing its tasks.
+	 */
+	uart_rx_disable(uart_dev);
+#endif
```
Comment (bjarki-trackunit) [inline tests/drivers/uart/uart_async_api/src/test_uart_async.c:600]: This is also an indication that the emulated UART needs to be changed to behave like a real UART. Use a delayable work item to simulate receiving and transmitting taking time :)

File: tests/drivers/uart/uart_async_api/src/test_uart.h (hunk 2 above)
Comment (bjarki-trackunit) [reply tests/drivers/uart/uart_async_api/src/test_uart.h:33]: This should not be needed, as the emulated device driver should behave like any other hardware device, if the test suite must be changed to support the emulated uart, it is not emulating a uart device properly :)

**Review by bjarki-andreasen — CHANGES_REQUESTED** [review]
Looks really promising, a few requests from my side

File: tests/drivers/uart/uart_emul/prj.conf (hunk 1 above)
Comment (dcpleung) [reply tests/drivers/uart/uart_emul/prj.conf:5]: I still think it is a good idea to test without enabling the API at all.

File: tests/drivers/uart/uart_async_api/src/test_uart_async.c (hunk 4 above)
Comment (rvoell-gcx) [reply tests/drivers/uart/uart_async_api/src/test_uart_async.c:600]: First of all, thank you for your review, i really appreciate this!

But I do not agree here. I feel that simulating the timing of each individual byte being sent or received is quite a bit of overkill and would add quite a bit of complexity to this. One could calculate when the events would arrive, but this feels even more complex. I think not having the correct timing here is acceptable, even when this means that the testcase needs a small change as a compromise.

When i think about, what i want to achieve as a developer when using this fixture, i usually want to test if my code did use the API correctly and does work with the events correctly. So the events need to come in in the correct order, have the correct payloads etc. Timing is not that big of an issue, i think. E.g. if one needs the code under test to receive a message at an exact point in time, one could simply provide the data with `uart_emul_put_rx_data()` at the correct moment in time.

Also the already existing code does not do that, the implementations of the polling and the interrupt based APIs do not respect accurate timing either.

File: drivers/serial/uart_emul.c [hunk 5]
```diff
@@ -48,6 +54,33 @@ struct uart_emul_data {
...
+
+#ifdef CONFIG_UART_ASYNC_API
+	bool rx_async_en;
+	bool rx_stopping;
+	bool rx_release_on_timeout;
+
+	struct uart_emul_work tx_work;
+	struct uart_emul_work rx_work;
+	struct uart_emul_work rx_disable_work;
+	struct uart_emul_delayable_work rx_timeout_work;
```
[the remark this answers is not in the document]
Comment (rvoell-gcx) [reply drivers/serial/uart_emul.c:66]: I adapted here to the "style" that was already present in the file. Is it okay if i fix this for the already existing code in this file, too? I would like to add a new commit for this before my current commits.

[the remark this answers is not in the document]
Comment (bjarki-trackunit) [reply drivers/serial/uart_emul.c:66]: I think it would be great if you updated the existing code in a commit as part of this PR :)

File: tests/drivers/uart/uart_async_api/src/test_uart_async.c (hunk 4 above)
Comment (bjarki-trackunit) [reply tests/drivers/uart/uart_async_api/src/test_uart_async.c:600]: That sounds more like a mock than an emulated UART, something better implemented using the FFF framework I think. Emulated device drivers are typically used to test the test suites themselves, and to provide a "fake" device that can be used to develop applications, like the emulated RTC and GNSS device drivers.

If the emulated UART is meant to be used as a mock, which it seems like it is, it should not be tested as part of the test suite for real UARTs, as it is not one (nor emulating one)

I would simply remove the emulated UART from the uart async api test suite, WDYT?

Comment (rvoell-gcx) [reply tests/drivers/uart/uart_async_api/src/test_uart_async.c:600]: Thats the other extreme. I would like to have it behave like a real UART  at the API level as much as possible and the already existing test cases come in handy to test this, as those test zephyrs expectation on how the API should behave. I also got to jump through some extra hoops to make it to comply with those test cases as much as possible. So i think its a bit more than a simple mock. I just don't wanna over engineer it.

File: tests/drivers/uart/uart_emul/prj.conf (hunk 1 above)
Comment (rvoell-gcx) [reply tests/drivers/uart/uart_emul/prj.conf]: I did split this into three test configurations

File: tests/drivers/uart/uart_async_api/src/test_uart.h (hunk 2 above)
Comment (rvoell-gcx) [reply tests/drivers/uart/uart_async_api/src/test_uart.h]: I removed this. The special case for the abort test is still there, though

File: drivers/serial/uart_emul.c (hunk 5 above)
[the remark this answers is not in the document]
Comment (rvoell-gcx) [reply drivers/serial/uart_emul.c]: I refactored this, hopefully its now prettier :)

File: drivers/serial/uart_emul.c [hunk 6]
```diff
@@ -331,6 +364,485 @@ static int uart_emul_irq_update(const struct device *dev)
...
+	}
+
+	uart_emul_post_event(dev, &(struct uart_event){
+					  .type = UART_RX_RDY,
+					  .data.rx = {
+							  .buf = rx_buf,
+							  .offset = rx_buf_offset,
+							  .len = rx_buf_data_len,
+						  },
+				  });
```
[the remark this answers is not in the document]
Comment (rvoell-gcx) [reply drivers/serial/uart_emul.c]: I fixed this

File: tests/drivers/uart/uart_async_api/testcase.yaml [hunk 7]
```diff
@@ -110,3 +110,8 @@ tests:
     extra_args:
       - DTC_OVERLAY_FILE="boards/nucleo_h723zg.overlay;boards/nucleo_h723zg_nocachemem.overlay"
       - EXTRA_CONF_FILE=stm32_nocache_mem_dt.conf
+  drivers.uart.async_api.uart_emul:
+    harness: ztest
+    platform_allow:
+      - qemu_x86
+      - native_posix
```
[the remark this answers is not in the document]
Comment (rvoell-gcx) [reply tests/drivers/uart/uart_async_api/testcase.yaml]: Using native_sim instead of native_posix now. I also moved the devicetree overlays into one special for the uart emulation node, which is then used for both platforms.

File: tests/drivers/uart/uart_async_api/src/test_uart_async.c (hunk 3 above)
Comment (rvoell-gcx) [reply tests/drivers/uart/uart_async_api/src/test_uart_async.c]: I removed this, was no longer necessary anyways.

File: tests/drivers/uart/uart_async_api/src/test_uart_async.c (hunk 4 above)
Comment (rvoell-gcx) [reply tests/drivers/uart/uart_async_api/src/test_uart_async.c]: Not sure how to resolve this. I would be quite unhappy with removing it from this tests and i also do not want add simulation of the timings here, as this would add quite a bit of complexity and needs a bit of work to do for unclear benefit.

Comment (bjarki-trackunit) [reply tests/drivers/uart/uart_async_api/src/test_uart_async.c]: The benefit is avoiding violating the UART API and test suite by coupling them to a specific device driver which is not behaving like a real UART, which is what the API and test suite are designed to abstract and validate.

Yes it is more complexity to make it behave like a real UART, but that in itself is not a sufficient argument against adding said complexity if the goal is to make it emulate a real UART. If the goal is not to emulate a real UART, don't add the additional complexity, and don't treat it like a real UART by for example adding it to the UART test suite :)

I'm not saying which is the better option, but those are the options AFAIK

File: tests/drivers/uart/uart_emul/src/main.c [hunk 8]
```diff
@@ -9,25 +9,30 @@
 #include <zephyr/drivers/serial/uart_emul.h>
 #include <zephyr/ztest.h>
 
-#define EMUL_UART_NODE	       DT_NODELABEL(euart0)
+#define EMUL_UART_NODE         DT_NODELABEL(euart0)
```
Comment (dcpleung) [inline tests/drivers/uart/uart_emul/src/main.c:12]: Please remove the whitespace change. This is not related.

File: tests/drivers/uart/uart_emul/src/main.c [hunk 9]
```diff
@@ -9,25 +9,30 @@
...
 #define EMUL_UART_RX_FIFO_SIZE DT_PROP(EMUL_UART_NODE, rx_fifo_size)
 #define EMUL_UART_TX_FIFO_SIZE DT_PROP(EMUL_UART_NODE, tx_fifo_size)
 
 /*
  * Leave one byte left in tx to avoid filling it completely which will block the UART
  * tx ready IRQ event.
  */
-#define SAMPLE_DATA_SIZE       MIN(EMUL_UART_RX_FIFO_SIZE, EMUL_UART_TX_FIFO_SIZE) - 1
+#define SAMPLE_DATA_SIZE MIN(EMUL_UART_RX_FIFO_SIZE, EMUL_UART_TX_FIFO_SIZE) - 1
```
Comment (dcpleung) [inline tests/drivers/uart/uart_emul/src/main.c:20]: Ditto on removing whitespace changes.

File: tests/drivers/uart/uart_emul/src/main.c (hunk 8 above)
Comment (rvoell-gcx) [reply tests/drivers/uart/uart_emul/src/main.c]: oops, i fixed that

File: tests/drivers/uart/uart_async_api/src/test_uart_async.c (hunk 4 above)
Comment (dcpleung) [reply tests/drivers/uart/uart_async_api/src/test_uart_async.c]: I do think it is better not to modify the tests to suit a particular driver.

File: tests/drivers/uart/uart_emul/testcase.yaml [hunk 10]
```diff
@@ -5,4 +5,22 @@ common:
...
+      - CONFIG_UART_INTERRUPT_DRIVEN=y
+    extra_args: DTC_OVERLAY_FILE="uart_emul.overlay"
+  drivers.uart.emul.async:
+    platform_allow:
+      - qemu_x86
+      - native_sim
+    extra_configs:
+      - CONFIG_EVENTS=y
+      - CONFIG_UART_ASYNC_API=y
+    extra_args: DTC_OVERLAY_FILE="uart_emul.overlay"
```
Comment (ajarmouni-st) [inline tests/drivers/uart/uart_emul/testcase.yaml:26]: You could move this under `common`.

Comment (ajarmouni-st) [reply tests/drivers/uart/uart_emul/testcase.yaml:26]: the same for
```
    platform_allow:
      - qemu_x86
      - native_sim
```

File: tests/drivers/uart/uart_async_api/src/test_uart_async.c (hunk 4 above)
Comment (rvoell-gcx) [reply tests/drivers/uart/uart_async_api/src/test_uart_async.c]: I dropped the commit enabling these tests now.

I'm not very happy about this, as i saw some value in validating the API as far as possible. But i do not have the time to write a complete simulation of a uart peripheral. I wrote this driver as part of a project to have a nice fixture for testing code using the uart async API and thought it might be useful for others, too. I think testing that the API does help to increase confidence in those tests.

File: tests/drivers/uart/uart_emul/testcase.yaml (hunk 10 above)
Comment (rvoell-gcx) [reply tests/drivers/uart/uart_emul/testcase.yaml]: Thank you for your review. I did that.

**Review by bjarki-andreasen — APPROVED** [review]
Nice work! You could extend the testcase to build for native_sim_64 and qemu_x86_64 to build and run it for 64-bit archs as well :)

**Review by cfriedt — APPROVED** [review]

**Review by dcpleung — APPROVED** [review]
